drivers/dma/dma_stm32.c

```diff
@@ -0,0 +1,616 @@
+	    config->dest_data_size != 2U &&
+	    config->dest_data_size != 1U) {
+		LOG_ERR("Dest unit size error, %d",
+			config->dest_data_size);
+		return -EINVAL;
+	}
+
+	/*
+	 * STM32's circular mode will auto reset both source address
+	 * counter and destination address counter.
+	 */
+	if (config->head_block->source_reload_en !=
+		config->head_block->dest_reload_en) {
+		LOG_ERR("source_reload_en and dest_reload_en must "
+			"be the same.");
+		return -EINVAL;
+	}
+
+	stream->busy		= true;
+	stream->dma_callback	= config->dma_callback;
+	stream->direction	= config->channel_direction;
+	stream->callback_arg    = config->callback_arg;
+
+	if (stream->direction == MEMORY_TO_PERIPHERAL) {
+		DMA_InitStruct.MemoryOrM2MDstAddress =
+					config->head_block->source_address;
+		DMA_InitStruct.PeriphOrM2MSrcAddress =
+					config->head_block->dest_address;
+	} else {
+		DMA_InitStruct.PeriphOrM2MSrcAddress =
+					config->head_block->source_address;
+		DMA_InitStruct.MemoryOrM2MDstAddress =
+					config->head_block->dest_address;
+	}
+
+	u16_t memory_addr_adj, periph_addr_adj;
+
+	ret = dma_stm32_get_priority(config->channel_priority,
+				     &DMA_InitStruct.Priority);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = dma_stm32_get_direction(config->channel_direction,
+				      &DMA_InitStruct.Direction);
+	if (ret < 0) {
+		return ret;
+	}
+
+	switch (config->channel_direction) {
+	case MEMORY_TO_MEMORY:
+	case PERIPHERAL_TO_MEMORY:
+		memory_addr_adj = config->head_block->dest_addr_adj;
+		periph_addr_adj = config->head_block->source_addr_adj;
+		break;
+	case MEMORY_TO_PERIPHERAL:
+		memory_addr_adj = config->head_block->source_addr_adj;
+		periph_addr_adj = config->head_block->dest_addr_adj;
+		break;
+	/* Direction has been asserted in dma_stm32_get_direction. */
+	}
+
+	ret = dma_stm32_get_memory_increment(memory_addr_adj,
+					&DMA_InitStruct.MemoryOrM2MDstIncMode);
+	if (ret < 0) {
+		return ret;
+	}
+	ret = dma_stm32_get_periph_increment(periph_addr_adj,
+					&DMA_InitStruct.PeriphOrM2MSrcIncMode);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (config->head_block->source_reload_en) {
+		DMA_InitStruct.Mode = LL_DMA_MODE_CIRCULAR;
+	} else {
+		DMA_InitStruct.Mode = LL_DMA_MODE_NORMAL;
+	}
+
+	source_periph = stream->direction == MEMORY_TO_PERIPHERAL;
+	ret = dma_stm32_width_config(config, source_periph, dma,
+				     &DMA_InitStruct, id);
+	if (ret < 0) {
+		return ret;
+	}
+	msize = DMA_InitStruct.MemoryOrM2MDstDataSize;
+
+#if defined(CONFIG_DMA_STM32_V1)
+	DMA_InitStruct.MemBurst = stm32_dma_get_mburst(config, source_periph);
+	DMA_InitStruct.PeriphBurst = stm32_dma_get_pburst(config,
+							source_periph);
+
+	if (config->channel_direction != MEMORY_TO_MEMORY) {
+		if (config->dma_slot >= 8) {
+			LOG_ERR("dma slot error.");
+			return -EINVAL;
+		}
+	} else {
+		if (config->dma_slot >= 8) {
+			LOG_ERR("dma slot is too big, using 0 as default.");
+			config->dma_slot = 0;
+		}
+	}
+	DMA_InitStruct.Channel = config->dma_slot;
```

Comment:
I think this is wrong.
DMA_InitStruct.Channel should be assigned with LL_DMA_CHANNEL_0/LL_DMA_CHANNEL_1/LL_DMA_CHANNEL_2/...

I noticed infact that the CR content was wrong here.
To fix it I added following code:

 >       switch (config->dma_slot) {
 >       case 0:
 >       >       DMA_InitStruct.Channel = LL_DMA_CHANNEL_0;
 >       >       break;
 >       case 1:
 >       >       DMA_InitStruct.Channel = LL_DMA_CHANNEL_1;
 >       >       break;
 >       case 2:
 >       >       DMA_InitStruct.Channel = LL_DMA_CHANNEL_2;
 >       >       break;
 [...]
 >      }